File: BasicsWithJava/WorkManagerJava/app/src/main/java/com/sumere/workmanagerjava/MainActivity.java
package com.sumere.workmanagerjava;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.work.Constraints;
import androidx.work.Data;
import androidx.work.NetworkType;
import androidx.work.OneTimeWorkRequest;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkInfo;
import androidx.work.WorkManager;
import androidx.work.WorkRequest;

import android.os.Bundle;

import java.util.concurrent.TimeUnit;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Data data = new Data.Builder().putInt("IncrementValue",1).build();
        /*
            Constraints constraints = new Constraints.Builder()
                .setRequiredNetworkType(NetworkType.CONNECTED)
                .build();
        */
        /*
        WorkRequest workRequest = new OneTimeWorkRequest.Builder(RefreshDatabase.class)
                //.setConstraints(constraints)
                .setInputData(data)
                //.setInitialDelay(5, TimeUnit.MINUTES)
                //.addTag("myTag")
                .build();
        WorkManager.getInstance(this).enqueue(workRequest);
        */
        WorkRequest workRequest = new PeriodicWorkRequest.Builder(RefreshDatabase.class,15,TimeUnit.MINUTES)
                .setInputData(data)
                .addTag("TempTag")
                .build();
        WorkManager.getInstance(this).enqueue(workRequest);
        WorkManager.getInstance(this).getWorkInfoByIdLiveData(workRequest.getId()).observe(this, new Observer<WorkInfo>() {
            @Override
            public void onChanged(WorkInfo workInfo) {
                if(workInfo.getState() == WorkInfo.State.RUNNING){
                    System.out.println("Running");
                }
            }
        });
        //WorkManager.getInstance(this).cancelAllWork();
        //WorkManager.getInstance(this).getWorkInfosByTag("TempTag");

        //Chaining

        /*
        WorkManager.getInstance(this).beginWith(OneTimeWorkRequest).
                then(OneTimeWorkRequest)
                .then(OneTimeWorkRequest).enqueue();
        */
    }

}

File: KotlinWorkManager/app/src/main/java/com/sumere/kotlinworkmanager/MainActivity.kt
package com.sumere.kotlinworkmanager

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.lifecycle.Observer
import androidx.work.Constraints
import androidx.work.Data
import androidx.work.NetworkType
import androidx.work.OneTimeWorkRequest
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.PeriodicWorkRequest
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkInfo
import androidx.work.WorkManager
import androidx.work.WorkRequest
import java.util.concurrent.TimeUnit

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val data = Data.Builder().putInt("intKey",1).build()
        val constraints = Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .setRequiresCharging(false)
            .build()


        val myWorkRequest: OneTimeWorkRequest = OneTimeWorkRequestBuilder<RefreshDatabase>()
            .setConstraints(constraints)
            .setInputData(data)
            //.setInitialDelay(15,TimeUnit.MINUTES)
            .build()
        //One Time Work Manager
        WorkManager.getInstance(this@MainActivity).enqueue(myWorkRequest)

        val myPeriodicWorkRequest: PeriodicWorkRequest = PeriodicWorkRequestBuilder<RefreshDatabase>(15,TimeUnit.MINUTES)
            .setConstraints(constraints)
            .setInputData(data)
            .build()
        //Periodic Work Manager
        WorkManager.getInstance(this@MainActivity).enqueue(myPeriodicWorkRequest)

        WorkManager.getInstance(this@MainActivity).getWorkInfoByIdLiveData(myPeriodicWorkRequest.id).observe(this@MainActivity,
            Observer {
                if (it.state == WorkInfo.State.RUNNING){
                    println("Running")
                } else if(it.state == WorkInfo.State.FAILED){
                    println("Failed")
                } else if(it.state == WorkInfo.State.SUCCEEDED){
                    println("Succeeded")
                }
            })

        //Chaining
        WorkManager.getInstance(this@MainActivity)
            .beginWith(myWorkRequest)
            .then(myWorkRequest)
            .then(myWorkRequest)
            .enqueue()

    }
}